File: task-back-end/src/test/java/be/stijnhooft/task/backend/task/mother/TaskMother.java
package be.stijnhooft.task.backend.task.mother;

import be.stijnhooft.task.backend.TestClock;
import be.stijnhooft.task.backend.task.domain.Task;
import be.stijnhooft.task.backend.task.domain.TaskPatch;
import be.stijnhooft.task.backend.task.domain.TaskStatus;
import org.jspecify.annotations.Nullable;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.temporal.ChronoUnit;
import java.util.UUID;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicLong;

/// Tasks built the way the application builds them: through the creation patch and the fold.
///
/// Instancio is deliberately not used here any more. A task assembled field by field is a task
/// whose columns do not agree with its own history, which is a state the application cannot reach
/// and a test asserting on it proves nothing. It also minted dates decades ahead, which #10 banned
/// after a stream test received a patch dated 2071 from another test class.
public class TaskMother {

    private static final TestClock CLOCK = TestClock.atNoonOn(LocalDate.of(2026, 3, 1));

    /// **Negative**, and that is the point. `sequence` is unique in the schema and the integration
    /// tests share one reused Postgres, so these numbers have to miss the real sequence - but they
    /// must also never *pose* as it. Minted above it, as they were, they became the highest sequence
    /// in the database, so `TaskPatchSequence.watermark()` reported a cursor a thousand times past
    /// the end of history and every stream resumed from a point no real patch will reach for years.
    /// The server only ever issues positive numbers, so a negative one is unmistakably test data.
    private static final AtomicLong SEQUENCES =
            new AtomicLong(ThreadLocalRandom.current().nextLong(-1_000_000_000_000L, -1_000_000L));

    /// Long enough that a query reading the closing patch's date instead of the firing date cannot
    /// accidentally agree with one that reads the right one.
    private static final int CLOSED_AFTER_DAYS = 20;

    /// A task with a creation patch and two later patches, all uniquely identified, so a test can
    /// assert on its own data by id.
    public static Task createRandomTask() {
        var task = Task.builderForInitialTask(CLOCK)
                .name("task " + UUID.randomUUID())
                .context("context " + UUID.randomUUID())
                .description("description " + UUID.randomUUID())
                .dueDate(LocalDate.of(2026, 3, 8))
                .build();

        return task
                .patch(patchOn(task.id(), 1, "name", "renamed " + UUID.randomUUID()))
                .patch(patchOn(task.id(), 2, "dueDate", "2026-03-09"))
                .withSequencesFrom(SEQUENCES::getAndIncrement);
    }

    /// A task a template fired, in the state it ended up in - the shape `TaskOccurrences` is asked
    /// about.
    ///
    /// The firing date **is** the creation date, so it is set rather than defaulted, and the closing
    /// patch is dated well after it: what the port answers with must be the date the template came
    /// round, never the day someone got to it.
    ///
    /// @param completedOn  the day the work happened, for a completion; null for a cancellation,
    ///                     which is a closure and not a completion
    public static Task firedTask(UUID templateId, LocalDate firingDate, TaskStatus status,
                                 @Nullable LocalDate completedOn) {
        var task = Task.builderForInitialTask(CLOCK)
                .name("task " + UUID.randomUUID())
                .context("context " + UUID.randomUUID())
                .creationDateTime(firingDate.atTime(LocalTime.NOON).atZone(TestClock.ZONE).toInstant())
                .startDate(firingDate)
                .taskTemplateId(templateId)
                .occurrenceId(UUID.randomUUID())
                .build();

        if (status != TaskStatus.OPEN) {
            var closing = TaskPatch.builder()
                    .taskId(task.id())
                    .dateTime(task.creationDateTime().plus(CLOSED_AFTER_DAYS, ChronoUnit.DAYS))
                    .change("status", status)
                    .change("completedOn", completedOn)
                    .build();
            task = task.patch(closing);
        }

        return task.withSequencesFrom(SEQUENCES::getAndIncrement);
    }

    private static TaskPatch patchOn(UUID taskId, int daysAfterCreation, String field, String value) {
        return TaskPatch.builder()
                .taskId(taskId)
                .dateTime(CLOCK.instant().plus(daysAfterCreation, ChronoUnit.DAYS))
                .change(field, value)
                .build();
    }
}

File: task-front-end/src/app/domain/task.mother.ts
import { IsoDate } from './dates';
import { Importance, Task, TaskPatch, TaskStatus } from './task';

/**
 * A task builder for tests. **Imported only by `*.spec.ts`** — nothing in `src/main` reaches it.
 *
 * It builds the *folded* shape rather than a history, because everything on the overview reads a
 * folded task: the ranking, the bands and the panel all take `Task` and none of them has ever seen
 * a patch. Where a test needs a real history — undo, closure dates — it builds the patches itself,
 * as `local-store.spec.ts` does.
 */

let minted = 0;

export interface TaskOverrides {
  readonly id?: string;
  readonly name?: string;
  readonly creationDateTime?: string;
  readonly startDate?: IsoDate;
  readonly dueDate?: IsoDate | null;
  readonly context?: string;
  readonly importance?: Importance;
  readonly description?: string | null;
  readonly status?: TaskStatus;
  readonly completedOn?: IsoDate | null;
  readonly cancelledOn?: IsoDate | null;
  readonly taskTemplateId?: string | null;
  readonly occurrenceId?: string | null;
  readonly history?: readonly TaskPatch[];
}

export function aTask(overrides: TaskOverrides = {}): Task {
  const id = overrides.id ?? `task-${++minted}`;
  const creationDateTime = overrides.creationDateTime ?? '2026-01-01T09:00:00Z';
  const task = {
    id,
    name: overrides.name ?? `Task ${id}`,
    creationDateTime,
    // Started by default: a test about the bands says so explicitly, and every other test wants a
    // task that is simply on screen.
    startDate: overrides.startDate ?? '2020-01-01',
    dueDate: overrides.dueDate ?? null,
    context: overrides.context ?? 'house',
    importance: overrides.importance ?? 'IMPORTANT',
    description: overrides.description ?? null,
    status: overrides.status ?? 'OPEN',
    completedOn: overrides.completedOn ?? null,
    cancelledOn: overrides.cancelledOn ?? null,
    taskTemplateId: overrides.taskTemplateId ?? null,
    occurrenceId: overrides.occurrenceId ?? null,
    history: [],
  } satisfies Task;

  return {
    ...task,
    // **A creation patch carries every field the fold requires**, because the fold replays from it
    // and a history missing one produces no task at all. Building it from the task above rather
    // than beside it is what keeps the two from disagreeing — a test that re-folds this history
    // must get the same task back.
    history: overrides.history ?? [
      {
        id: `${id}-creation`,
        taskId: id,
        dateTime: creationDateTime,
        sequence: null,
        voids: null,
        changes: {
          name: task.name,
          creationDateTime: task.creationDateTime,
          startDate: task.startDate,
          dueDate: task.dueDate,
          context: task.context,
          importance: task.importance,
          description: task.description,
          status: task.status,
          completedOn: task.completedOn,
          taskTemplateId: task.taskTemplateId,
          occurrenceId: task.occurrenceId,
        },
      },
    ],
  };
}
